### ISSUE TITLE
sprint production release 2021.07.05

### ISSUE THREAD

**Issue by harshad16** [thoth-station/thoth-application#1757]
Hello, Thoth-station!

This Issue would be used for the current sprint cycle production release.
By the end of the sprint cycle, we will consolidate the information of thoth-station components features upgrade and fixes in this issue.

**Comment by fridex** [comment]
**Resolve Python software stacks based on RPM packages present in the base container image (runtime environment used)**

Some Python libraries require native packages to work. An example can be [GitPython](https://pypi.org/project/GitPython/), a popular Python library for interacting with Git. The installation of GitPython done by `pip install gitpython` will bring Python library into the environment, nevertheless, the library does require Git to be installed in the environment to work properly. The newly introduced mechanism allows to declaratively specify which RPM packages (such as [Git required by GitPython in the example](https://github.com/thoth-station/prescriptions/blob/01c6fcac25772175c864b365503e5a802658694a/prescriptions/gitpython/rpm.yaml)) should be present in the environment so that the resolved software will serve user needs. See [documentation for more info](https://thoth-station.ninja/docs/developers/adviser/prescription.html#should-include-runtime-environments-rpm-packages).

* https://github.com/thoth-station/adviser/issues/1922

**Comment by fridex** [comment]
**Resolve Python software stacks based on shared objects present in the base container image (runtime environment used)**

Similarly as for RPM packages (see above), some Python packages require native shared objects (`so` files) present in the environment to operate correctly. To support this, Thoth performs analyses of base container images that can be used with Python software stacks. One of the outcomes of these analyses is information about shared objects (and their versions) shipped in the container images. Detection of required shared objects in Python packages then creates a requirement to match shared objects offered by the base container image used (done by a [corresponding ABI sieve](https://thoth-station.ninja/docs/developers/adviser/thoth.adviser.sieves.html?)).

Besides this automatic detection, Thoth offers a declarative interface to describe pipeline units which adjust the resolution process based on shared objects present in the base container image. This is especially useful if shared objects are loaded dynamically on runtime - maintainers of such Python packages can state requirements on the base container image with respect to shared objects and their versions present. An example can be a requirement on a specific glibc version - see [docs for more information](https://thoth-station.ninja/docs/developers/adviser/prescription.html#should-include-runtime-environments-shared-objects).

* https://github.com/thoth-station/adviser/issues/1925

**Comment by fridex** [comment]
**Pulp integration: pulp-pypi-sync-job**

[Pulp](https://pulpproject.org/) offers [pulp-python](https://pulp-python.readthedocs.io/en/latest/) plugin which enables users to maintain multiple Python package indexes. Users can upload their Python packages to package indexes considering their needs (e.g. have a specific Python package index for AVX2 optimized TensorFlow builds, another index for patched Python packages needed for testing products, ...). To integrate with Thoth, we introduced [pulp-pypi-sync-job](https://github.com/thoth-station/pulp-pypi-sync-job) which periodically checks Python package indexes created on a monitored Pulp instance and registers them to Thoth. Subsequently, Thoth analyses packages published on these package indexes and can guide users on how to consume packages published there. As Thoth offers resolution which, unlike pip/Pipenv, considers Python package indexes as a different source of packages (rather than mirrors as in the case of pip/Pipenv), the resolution process can give guidance on how and why Python packages should be consumed from the recommended Python package sources.

As of now, we monitor [Pulp instance deployed on the operate-first environment](https://github.com/operate-first/support/issues/176).

* https://github.com/thoth-station/thoth-application/issues/1733

**Comment by goern** [comment]
Updated Knowledge Graph: https://github.com/thoth-station/core/issues/297

**Comment by goern** [comment]
sorry for the verbose testing :/

**Comment by goern** [comment]
looks like we need to create a milestone manually and then use a comment-command
/milestone 2021.07.05

**Comment by goern** [comment]
https://github.com/thoth-station/thamos/pull/886

**Comment by goern** [comment]
https://github.com/thoth-station/package-extract/pull/452

**Comment by fridex** [comment]
> [thoth-station/package-extract#452](https://github.com/thoth-station/package-extract/pull/452)

I would postpone this one for the next release - the change is in one component, further adjustments in other components are needed to deliver a usable feature - https://github.com/thoth-station/adviser/issues/1926

**Comment by goern** [comment]
https://github.com/thoth-station/ps-nlp/issues/10#issuecomment-872764700

**Comment by harshad16** [comment]
we have completed the release of 2021.07.05 :tada: :confetti_ball: :partying_face: 

### Features

* **Resolve Python software stacks based on RPM packages present in the base container image (runtime environment used)**

Some Python libraries require native packages to work. An example can be [GitPython](https://pypi.org/project/GitPython/), a popular Python library for interacting with Git. The installation of GitPython done by `pip install gitpython` will bring Python library into the environment, nevertheless, the library does require Git to be installed in the environment to work properly. The newly introduced mechanism allows to declaratively specify which RPM packages (such as [Git required by GitPython in the example](https://github.com/thoth-station/prescriptions/blob/01c6fcac25772175c864b365503e5a802658694a/prescriptions/gitpython/rpm.yaml)) should be present in the environment so that the resolved software will serve user needs. See [documentation for more info](https://thoth-station.ninja/docs/developers/adviser/prescription.html#should-include-runtime-environments-rpm-packages).

* https://github.com/thoth-station/adviser/issues/1922

* **Resolve Python software stacks based on shared objects present in the base container image (runtime environment used)**

Similarly as for RPM packages (see above), some Python packages require native shared objects (`so` files) present in the environment to operate correctly. To support this, Thoth performs analyses of base container images that can be used with Python software stacks. One of the outcomes of these analyses is information about shared objects (and their versions) shipped in the container images. Detection of required shared objects in Python packages then creates a requirement to match shared objects offered by the base container image used (done by a [corresponding ABI sieve](https://thoth-station.ninja/docs/developers/adviser/thoth.adviser.sieves.html?#module-thoth.adviser.sieves.thoth_s2i_abi_compat)).

Besides this automatic detection, Thoth offers a declarative interface to describe pipeline units which adjust the resolution process based on shared objects present in the base container image. This is especially useful if shared objects are loaded dynamically on runtime - maintainers of such Python packages can state requirements on the base container image with respect to shared objects and their versions present. An example can be a requirement on a specific glibc version - see [docs for more information](https://thoth-station.ninja/docs/developers/adviser/prescription.html#should-include-runtime-environments-shared-objects).

* https://github.com/thoth-station/adviser/issues/1925

* **Pulp integration: pulp-pypi-sync-job**

[Pulp](https://pulpproject.org/) offers [pulp-python](https://pulp-python.readthedocs.io/en/latest/) plugin which enables users to maintain multiple Python package indexes. Users can upload their Python packages to package indexes considering their needs (e.g. have a specific Python package index for AVX2 optimized TensorFlow builds, another index for patched Python packages needed for testing products, ...). To integrate with Thoth, we introduced [pulp-pypi-sync-job](https://github.com/thoth-station/pulp-pypi-sync-job) which periodically checks Python package indexes created on a monitored Pulp instance and registers them to Thoth. Subsequently, Thoth analyses packages published on these package indexes and can guide users on how to consume packages published there. As Thoth offers resolution which, unlike pip/Pipenv, considers Python package indexes as a different source of packages (rather than mirrors as in the case of pip/Pipenv), the resolution process can give guidance on how and why Python packages should be consumed from the recommended Python package sources.

As of now, we monitor [Pulp instance deployed on the operate-first environment](https://github.com/operate-first/support/issues/176).

* https://github.com/thoth-station/thoth-application/issues/1733 

*  Initiate to provide base image for the natural language processing , we have created the following image for use:
https://quay.io/repository/thoth-station/ps-nlp?tab=tags
https://quay.io/repository/thoth-station/ps-nlp-tensorflow?tab=tags
https://quay.io/repository/thoth-station/ps-nlp-pytorch?tab=tags

* Introduced a new component which periodically exports metrics out of the main database asynchronously.

* Project Thoth is recognizing  GitHub published [GHSA-4278-2v5v-65r4](https://github.com/advisories/GHSA-4278-2v5v-65r4) which affects some Tensorflow version as "CVE" and advise accordingly.

### Component Updates

- [x] adviser: v0.35.0
  - Point users to support repository to report issues (https://github.com/thoth-station/adviser/pull/1949)
  - Keep pipeline run information only when adviser is run in verbose mode (https://github.com/thoth-station/adviser/pull/1944)
  - Add demos from sprint 63 (https://github.com/thoth-station/adviser/pull/1942)
  - Adjust documentation for prescriptions living in a dir structure (https://github.com/thoth-station/adviser/pull/1938)
  - Extend documentation for security (https://github.com/thoth-station/adviser/pull/1934)
  - Adjust reporting of RPM and SO specific issues (https://github.com/thoth-station/adviser/pull/1933)
  - Perform recommendations based on RPM packages present in the environment (https://github.com/thoth-station/adviser/pull/1928)
  - Adjust validate logic to consume prescriptions from a dir (https://github.com/thoth-station/adviser/pull/1929)
  - Filter Python packages present in software stack based on so files (https://github.com/thoth-station/adviser/pull/1927)

- [x] metrics-exporter: v0.16.3
  - Use common enums (https://github.com/thoth-station/metrics-exporter/pull/733)
  - Add missing labels (https://github.com/thoth-station/metrics-exporter/pull/730)
  - missin definition of metric (https://github.com/thoth-station/metrics-exporter/pull/719)

- [x] slo-reporter: v0.15.0
  - Add env label (https://github.com/thoth-station/slo-reporter/pull/278)
  - Remove qeb-hwt bits (https://github.com/thoth-station/slo-reporter/pull/283)

Thanks for the amazing work everyone. :100:
